### PR TITLE
refactor(build): Better encapsulate the broccoli builder.

### PR DESCRIPTION
**Pull request by alexeagle**

### CODE REVIEW

File: tools/broccoli/trees/common_js_tree.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+'use strict';
```
Comment (IgorMinar) [inline tools/broccoli/trees/common_js_tree.ts]: can you call this file `node` rather than `common`?

File: tools/broccoli/broccoli_builder.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+export class BroccoliBuilder {
+  private builder;
+  private broccoliExecuted = {};
+
+  // Named constructors
+  static forDevTree(): BroccoliBuilder {
+    return new BroccoliBuilder(makeBrowserTree({name: 'dev', typeAssertions: true}),
+                               path.join('js', 'dev'));
+  }
+  static forCjsTree(): BroccoliBuilder {
```
Comment (IgorMinar) [inline tools/broccoli/broccoli_builder.ts]: `forNoteTree`

File: gulpfile.js [hunk 3]
```diff
@@ -36,7 +36,7 @@ var insert = require('gulp-insert');
 function missingDynamicBroccoli() {
   throw new Error('ERROR: build.broccoli.tools task should have been run before using broccoli');
 }
-var makeBroccoliTree = missingDynamicBroccoli, broccoliBuild = missingDynamicBroccoli;
+var BroccoliBuilder = missingDynamicBroccoli;
```
Comment (mprobst) [inline gulpfile.js]: calling .forDartTree will now just throw undefined is not a function, your trick no longer works.

What about making this a function "getBroccoli"? It's a bit more transparent in any case.

File: tools/broccoli/broccoli_builder.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+var makeDartTree = require('./trees/dart_tree');
+var path = require('path');
+var printSlowTrees = require('broccoli-slow-trees');
+
+/**
+ * Helper for building with broccoli.
+ */
+export class BroccoliBuilder {
+  private builder;
+  private broccoliExecuted = {};
```
Comment (mprobst) [inline tools/broccoli/broccoli_builder.ts]: types?

Comment (alexeagle) [inline tools/broccoli/broccoli_builder.ts]: Types would be good, but we need to write more typings for broccoli.

File: tools/broccoli/broccoli_builder.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  constructor(tree, private outputRoot: string) {
+    if (this.broccoliExecuted.hasOwnProperty(outputRoot)) {
+      throw new Error('The broccoli task can be called only once for outputRoot ' + outputRoot);
+    }
+    this.broccoliExecuted[outputRoot] = true;
+
+    var distPath = path.join('dist', outputRoot);
+
+    // We do a clean build every time to avoid stale outputs.
```
Comment (mprobst) [inline tools/broccoli/broccoli_builder.ts]: So broccoli doesn't clean up the output folder if it already exists?

File: tools/broccoli/broccoli_builder.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+          // Should show file and line/col if present
+          if (err.file) {
+            console.error('File: ' + err.file);
+          }
+          if (err.stack) {
+            console.error(err.stack);
+          } else {
+            console.error(err);
+          }
+          throw err;
```
Comment (mprobst) [inline tools/broccoli/broccoli_builder.ts]: log and rethrow is usually an anti pattern - shouldn't the outer error handler already do that?

File: tools/broccoli/broccoli_builder.ts (hunk 5 above)
Comment (alexeagle) [inline tools/broccoli/broccoli_builder.ts]: default behavior in broccoli is to error out if the output folder exists

Comment (alexeagle) [inline tools/broccoli/broccoli_builder.ts]: We want to share the error-printing between run once and watch mode. But in
run once mode we also want to exit(1)

File: tools/broccoli/broccoli_builder.ts (hunk 4 above)
Comment (mprobst) [inline tools/broccoli/broccoli_builder.ts]: Yes, but broccoliExecuted is {[s: string]: boolean} (I called that Set somewhere else).

**Pull request merged**